pkg/generator/generator.go

```diff
@@ -16,32 +16,87 @@ package generator
 	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
-	"k8s.io/apimachinery/pkg/util/runtime"
+	runtimeutils "k8s.io/apimachinery/pkg/util/runtime"
 )
 
-var cmd = "/usr/bin/env bash %s"
-var cloudInitGenerator *ostemplate.CloudInitGenerator
-
-func additionalValues(*extensionsv1alpha1.OperatingSystemConfig) (map[string]interface{}, error) {
-	return nil, nil
-}
+var (
+	cmd                = "/usr/bin/env bash %s"
+	cloudInitGenerator *GardenLinuxCloudInitGenerator
+	unitsToEnable      []string
+)
 
 //go:embed templates/*
 var templates embed.FS
 
+// GardenLinuxCloudInitGenerator is a wrapper around the CloudInitGenerator from oscommon that implements its own generate function
+type GardenLinuxCloudInitGenerator struct {
+	cloudInitGenerator *ostemplate.CloudInitGenerator
+}
+
 func init() {
-	cloudInitTemplateString, err := templates.ReadFile("templates/cloud-init.gardenlinux.template")
-	runtime.Must(err)
+	cloudInitTemplateString, err := templates.ReadFile(filepath.Join("templates", "cloud-init.gardenlinux.template"))
+	runtimeutils.Must(err)
 
 	cloudInitTemplate, err := ostemplate.NewTemplate("cloud-init").Parse(string(cloudInitTemplateString))
-	runtime.Must(err)
-	cloudInitGenerator = ostemplate.NewCloudInitGenerator(cloudInitTemplate, ostemplate.DefaultUnitsPath, cmd, additionalValues)
+	runtimeutils.Must(err)
+
+	cloudInitGenerator = &GardenLinuxCloudInitGenerator{
+		cloudInitGenerator: ostemplate.NewCloudInitGenerator(cloudInitTemplate, ostemplate.DefaultUnitsPath, cmd, additionalValues),
+	}
+}
+
+// additionalValues provides additional values to the cloud-init template
+func additionalValues(*extensionsv1alpha1.OperatingSystemConfig) (map[string]interface{}, error) {
+	return map[string]interface{}{
+		"unitsToEnable": unitsToEnable,
+	}, nil
+}
+
+// Generate generates a Garden Linux specific cloud-init script from the given OperatingSystemConfig.
+func (g *GardenLinuxCloudInitGenerator) Generate(logger logr.Logger, osc *generator.OperatingSystemConfig) ([]byte, *string, error) {
+
+	if osc.Object.Spec.Type == gardenlinux.OSTypeGardenLinux &&
+		osc.Object.Spec.Purpose == extensionsv1alpha1.OperatingSystemConfigPurposeReconcile {
+
+		// add additional scripts that are provided in the embedded fs
+		additionalScripts, err := gardenlinux.GetAdditionalScripts()
+		if err != nil {
+			return nil, nil, err
+		}
+		osc.Files = append(osc.Files, additionalScripts...)
+
+		// add scripts and dropins for kubelet
+		kubeletCgroupdriverScript, kubeletCgroupdriverDropin, err := gardenlinux.KubeletCgroupDriver()
+		if err != nil {
+			return nil, nil, err
+		}
+		osc.Files = append(osc.Files, kubeletCgroupdriverScript...)
+		osc.Units = append(osc.Units, kubeletCgroupdriverDropin...)
+
+		// add scripts and dropins for containerd if activated
+		if osc.CRI != nil && osc.CRI.Name == extensionsv1alpha1.CRINameContainerD {
```

Comment:
I think we still maintain docker support. What happens if GL uses cgroupsv2 and the Shoot uses docker as container runtime? Would that lead to the kubelet having systemd, and docker having `cgroupfs` configured?

Reply:
Since Docker support has been removed from upstream Gardener and the amount of clusters still using Docker should be on the vanishing side, we are not touching worker pools without containerd support. 

But thanks for pointing this out, I accidentally reconfigured the kubelet regardless of the container runtime which would have led to problems on nodes with Docker as container runtime. Now, cgroup v2 enabled nodes with Docker will not get reconfigured and stay on `cgroupfs` (for both, kubelet and dockerd).